### ISSUE TITLE
onCompleted() if nothing found at a specified key.

### ISSUE THREAD

**Issue by dsvoronin** [Naturally-Being/G-Rex#6]
Don't you think it will be more natural if grex getters will fall into error while can't find value for key?
I see popular scenario here: i'am trying to get something persisted, and if i can't i'll go fetch from web.
Rx have some useful methods to handle error state, but emitting null not so cool.

example: 

What i'm doing now:
```
 persister.get(key, City.class)
        .filter(city -> {
          if (city == null) {
            throw new NullPointerException("Can't get persisted value for: " + key);
          }
          return true;
        })
        .onErrorResumeNext(
            locationController.detectNearestCity().doOnNext(this::showDetectedCityInfo))
        .subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(cityProperty::set, error -> Timber.d(error, "can't get selected city property"));
```

and with emitting error:
```
 persister.get(key, City.class)
        .onErrorResumeNext(locationController.detectNearestCity().doOnNext(this::showDetectedCityInfo))
        .subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(cityProperty::set, error -> Timber.d(error, "can't get selected city property"));
```

**Comment by chris-horner** [comment]
I feel like a neater way might be to immediately call `onComplete()` from inside Grex if nothing is found at the key, but from what I can tell, there's no `onEmptyResumeNext()` in RxJava. (And perhaps there won't be any time soon https://github.com/ReactiveX/RxJava/issues/681).

I totally agree with you, returning null doesn't feel quite right. Worse, the behaviour is inconsistent between requesting a single object and requesting a list. I'll jump on this and make the change to throwing an error as soon as I can :)

**Comment by grodin** [comment]
Before you do, please reconsider. I totally agree that you should not be returning `null` but I don't agree that `onError()` is the right way. I really think that if nothing is found (either because the file doesn't exist or the value isn't in the file) you should return an empty `Observable`, i.e. call `onComplete` immediately.

This matches the reactive style much more and is consistent with `Observable#from()` for example. Any downstream processing just won't do anything.

As to your final point, since 1.05, RxJava has `switchIfEmpty()` which is exactly what you're talking about. It's not hard to write in any case, and I think it makes the API more messy if not finding a requested value is an error. In particular, it means the API is leaking the implementation detail that it uses a file.

I thought of opening a PR to change exactly this a while ago, but it breaks the API so I had assumed it wouldn't be useful.

**Comment by dsvoronin** [comment]
Agree! Null was wrong and i suggest some better solution, but not so good as "empty".
API break never a reason to drop correctness change!

This code works same as first ones:
```
 persister.get(key, City.class)
        .filter(city -> city != null)
        .switchIfEmpty(locationController.detectNearestCity().doOnNext(this::showDetectedCityInfo))
```

**Comment by chris-horner** [comment]
Ah! I had no idea about `switchIfEmpty()`. That's fantastic! I'll definitely update both object and list methods to complete immediately if they find nothing.

Thanks for the tip :)

**Comment by chris-horner** [comment]
Fixed in 36cb07e81d0654de36c2669841fd8b0ab1301130

G-Rex 2.0.0 now pushed to Central. Thanks for the help and discussion, guys!
